src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// Used to read an 'application/x-www-form-urlencoded' form.
+    /// Internally reads from a PipeReader.
+    /// </summary>
+    public class FormPipeReader
+    {
+        private static int DefaultValueCountLimit = 1024;
+        private static int DefaultKeyLengthLimit = 1024 * 2;
+        private static int DefaultValueLengthLimit = 1024 * 1024 * 4;
+        private static ReadOnlySpan<byte> UTF8EqualEncoded => new byte[] { (byte)'=' };
+        private static ReadOnlySpan<byte> UTF8AndEncoded => new byte[] { (byte)'&' };
```

Comment:
We can hardcode utf16 but I dunno how common it is.

Reply:
Don't bother until you actually find some examples. I haven't seen it used here.

Reply:
Good

Reply:
We don't optimize Unicode (still call Uri.Unescape) so I'm not going to bother.